checker/src/org/checkerframework/checker/optional/OptionalVisitor.java

```diff
@@ -0,0 +1,310 @@
+                "java.util.Optional", methodName, params, atypeFactory.getProcessingEnv());
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.get */
+    private boolean isCallToGet(ExpressionTree expression) {
+        return TreeUtils.isMethodInvocation(expression, getMethod, atypeFactory.getProcessingEnv());
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.isPresent */
+    private boolean isCallToIsPresent(ExpressionTree expression) {
+        return TreeUtils.isMethodInvocation(
+                expression, isPresentMethod, atypeFactory.getProcessingEnv());
+    }
+
+    // Optional creation: of, ofNullable.
+    private boolean isOptionalCreation(MethodInvocationTree methInvok) {
+        ExecutableElement invoked = TreeUtils.elementFromUse(methInvok);
+        ProcessingEnvironment env = atypeFactory.getProcessingEnv();
+        return ElementUtils.isMethod(invoked, ofMethod, env)
+                || ElementUtils.isMethod(invoked, ofNullableMethod, env);
+    }
+
+    // Optional elimination: get, orElse, orElseGet, orElseThrow.
+    private boolean isOptionalElimation(MethodInvocationTree methInvok) {
+        ExecutableElement invoked = TreeUtils.elementFromUse(methInvok);
+        ProcessingEnvironment env = atypeFactory.getProcessingEnv();
+        return ElementUtils.isMethod(invoked, getMethod, env)
+                || ElementUtils.isMethod(invoked, orElseMethod, env)
+                || ElementUtils.isMethod(invoked, orElseGetMethod, env)
+                || ElementUtils.isMethod(invoked, orElseThrowMethod, env);
+    }
+
+    @Override
+    public Void visitConditionalExpression(ConditionalExpressionTree node, Void p) {
+        handleTernaryIsPresentGet(node);
+        return super.visitConditionalExpression(node, p);
+    }
+
+    /**
+     * Part of rule #3.
+     *
+     * <p>Pattern match for: {@code VAR.isPresent() ? VAR.get().METHOD() : VALUE}
+     *
+     * <p>Prefer: {@code VAR.map(METHOD).orElse(VALUE);}
+     */
+    public void handleTernaryIsPresentGet(ConditionalExpressionTree node) {
+
+        ExpressionTree condExpr = TreeUtils.skipParens(node.getCondition());
+        ExpressionTree trueExpr = TreeUtils.skipParens(node.getTrueExpression());
+        ExpressionTree falseExpr = TreeUtils.skipParens(node.getFalseExpression());
+
+        if (!isCallToIsPresent(condExpr)) {
+            return;
+        }
+        ExpressionTree receiver = TreeUtils.getReceiverTree(condExpr);
+
+        if (trueExpr.getKind() != Kind.METHOD_INVOCATION) {
+            return;
+        }
+        ExpressionTree trueReceiver = TreeUtils.getReceiverTree(trueExpr);
+        if (!isCallToGet(trueReceiver)) {
+            return;
+        }
+        ExpressionTree getReceiver = TreeUtils.getReceiverTree(trueReceiver);
+
+        // What is a better way to do this than string comparison?
+        if (receiver.toString().equals(getReceiver.toString())) {
```

Comment:
For some trees, you could convert to`FlowExpression.Receiver` and use their equals method.

To convert:
`org.checkerframework.dataflow.analysis.FlowExpressions#internalReprOf(AnnotationProvider, ExpressionTree)`

(An AnnotatedTypeFactory is an AnnotationProvider)

Before using equals, check that the receivers aren't null and that they don't `containsUnknown()`.